x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ClusterStateWaitUntilThresholdStep.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.xpack.core.ilm.step.info.SingleMessageFieldInfo;
+
+import java.time.Clock;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.xpack.core.ilm.LifecycleExecutionState.fromIndexMetadata;
+
+/**
+ * This step wraps an {@link ClusterStateWaitStep} in order to be able to manipulate what the next step will be, depending on the result of
+ * the wrapped {@link ClusterStateWaitStep}.
+ * <p>
+ * If the action response is complete, the {@link ClusterStateWaitUntilThresholdStep}'s nextStepKey will be the nextStepKey of the
+ * wrapped action. When the threshold level is surpassed, if the underlying step's condition was not met, the nextStepKey will be changed to
+ * the provided {@link #nextKeyOnThresholdBreach} and this step will stop waiting.
+ *
+ * Failures encountered whilst executing the wrapped action will be propagated directly.
+ */
+public class ClusterStateWaitUntilThresholdStep extends ClusterStateWaitStep {
+
+    private static final Logger logger = LogManager.getLogger(ClusterStateWaitUntilThresholdStep.class);
+
+    private final ClusterStateWaitStep stepToExecute;
+    private final StepKey nextKeyOnThresholdBreach;
+    private final AtomicBoolean thresholdPassed = new AtomicBoolean(false);
+
+    public ClusterStateWaitUntilThresholdStep(ClusterStateWaitStep stepToExecute, StepKey nextKeyOnThresholdBreach) {
+        super(stepToExecute.getKey(), stepToExecute.getNextStepKey());
+        this.stepToExecute = stepToExecute;
+        this.nextKeyOnThresholdBreach = nextKeyOnThresholdBreach;
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    @Override
+    public Result isConditionMet(Index index, ClusterState clusterState) {
+        IndexMetadata idxMeta = clusterState.metadata().index(index);
+        if (idxMeta == null) {
+            // Index must have been since deleted, ignore it
+            logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists",
+                getKey().getAction(), index.getName());
+            return new Result(false, null);
+        }
+
+        Result stepResult = stepToExecute.isConditionMet(index, clusterState);
+
+        if (stepResult.isComplete() == false) {
+            // checking the threshold after we execute the step to make sure we execute the wrapped step at least once (because time is a
+            // wonderful thing)
+            TimeValue retryThreshold = LifecycleSettings.LIFECYCLE_STEP_WAIT_TIME_THRESHOLD_SETTING.get(idxMeta.getSettings());
+            LifecycleExecutionState lifecycleState = fromIndexMetadata(idxMeta);
+            if (waitedMoreThanThresholdLevel(retryThreshold, lifecycleState, Clock.systemUTC())) {
```

Comment:
Oh, you're absolutely right - wall clocks are not trustworthy. I don't think `nanoTime` would be the answer and the reading invariant on that api is to be monotonically increasing, more so than accuracy.

I don't think this is too bad because if we get a reading where the time drift is hours away we'll retry the shrink cycle (so nothing truly bad happens, except a wasteful - and presumbaly eventually successful - shrink operation, with the note that we'll at least delete the abandoned shrunk index)